mlir/lib/Conversion/LoSPNtoCPU/Vectorization/SLP/GraphConversion.cpp

```diff
@@ -0,0 +1,499 @@
+}
+
+bool ConversionState::alreadyComputed(Superword* superword) const {
+  if (temporaryData.computedSuperwords.count(superword)) {
+    return true;
+  }
+  return permanentData.computedSuperwords.count(superword);
+}
+
+bool ConversionState::alreadyComputed(Value value) const {
+  return permanentData.alreadyComputed(value) || temporaryData.alreadyComputed(value);
+}
+
+bool ConversionState::isExtractable(Value value) {
+  if (temporaryData.extractableScalarValues.count(value)) {
+    return true;
+  }
+  return permanentData.extractableScalarValues.count(value);
+}
+
+void ConversionState::markComputed(Value value) {
+  assert(!alreadyComputed(value) && "marking already computed value as computed");
+  if (temporaryData.computedScalarValues.insert(value).second) {
+    for (auto const& callback : scalarCallbacks) {
+      callback(value);
+    }
+    if (auto* definingOp = value.getDefiningOp()) {
+      for (auto operand : definingOp->getOperands()) {
+        if (!alreadyComputed(operand)) {
+          markComputed(operand);
+        }
+      }
+    }
+  }
+}
+
+void ConversionState::markComputed(Superword* superword, Value value) {
+  assert(!alreadyComputed(superword) && "superword has already been converted");
+  for (auto* operand : superword->getOperands()) {
+    assert(alreadyComputed(operand) && "computing vector before its operands");
+  }
+  temporaryData.computedSuperwords[superword] = value;
+  for (size_t lane = 0; lane < superword->numLanes(); ++lane) {
+    if (!superword->hasAlteredSemanticsInLane(lane)) {
+      temporaryData.extractableScalarValues.try_emplace(superword->getElement(lane), superword, lane);
+    }
+  }
+  for (auto const& callback : vectorCallbacks) {
+    callback(superword);
+  }
+}
+
+void ConversionState::markExtracted(Value value) {
+  assert(!alreadyComputed(value) && "extracting value that has been marked as computed already");
+  if (temporaryData.extractedScalarValues.insert(value).second) {
+    for (auto const& callback : extractionCallbacks) {
+      callback(value);
+    }
+  }
+}
+
+Value ConversionState::getValue(Superword* superword) const {
+  auto value = temporaryData.computedSuperwords.lookup(superword);
+  if (value) {
+    return value;
+  }
+  value = permanentData.computedSuperwords.lookup(superword);
+  assert(value && "the superword has not yet been converted");
+  return value;
+}
+
+ValuePosition ConversionState::getSuperwordContainingValue(Value value) const {
+  auto valuePosition = temporaryData.extractableScalarValues.lookup(value);
+  if (valuePosition) {
+    return valuePosition;
+  }
+  valuePosition = permanentData.extractableScalarValues.lookup(value);
+  assert(valuePosition && "there is no superword containing the value");
+  return valuePosition;
+}
+
+// Helper functions in anonymous namespace.
+namespace {
+  /// There is no isBeforeInBlock() for values in MLIR (only for Operation*), so we just define our own one.
+  bool isBeforeInBlock(Value lhs, Value rhs) {
```

Comment:
This function is used during the [deterministic computation of the conversion order](https://github.com/esa-tu-darmstadt/spn-compiler/blob/f087c57cddc2d863b1c27ea81901e939ae8aa28b/mlir/lib/Conversion/LoSPNtoCPU/Vectorization/SLP/GraphConversion.cpp#L184) and necessary for a determinstic superword order. It's not deprecated.

MLIR's `isBeforeInBlock(Operation*, Operation*)` also still appears here and there in the SLP vectorization. It should not appear during the actual pattern application process however, since that's where its performance problems actually appear (it struggles with basic block changes, but does not have any performance problems if a basic block remains unchanged in between calls).

I'm unsure why you think these methods are deprecated or not used anymore. Would you mind elaborating on that?

Reply:
I only remembered our discussion about the bad performance of `isBeforeInBlock` and found the use of the function only after adding this comment, so I thought it might be superfluous. Thanks for the explanation.